### PR TITLE
pubsys: enable the use of crane to publish kits

### LINKED ISSUE #316: Add crane support for publishing kits
We should add support for using crane for publishing kits if the tool is installed on the builder host

### PR DESCRIPTION
**Pull request by sam-berning**
**Issue number:**

Closes #316 

**Description of changes:**

Adds methods to `oci_cli_wrapper::ImageTool` for pushing individual-arch kit images and the multi-arch manifest, and implements these methods for both docker and crane.

Additionally, moves the `DockerArchitecture` enum into `oci-cli-wrapper`, since it is mainly used while interacting with the OCI CLI tools.

Finally, cleaned up unused dependencies across Twoliter with `cargo-machete` (barring all of the `bindeps` dependencies for Twoliter)

**Testing done:**

Built twoliter, and tried:

```
$ ../../../target/debug/pubsys --infra-config-path Infra.toml publish-kit --kit-path build/kits/core-kit --vendor bottlerocket --version v1.0.0 --build-id c232d4f-dirty
```

```
$ TWOLITER_KIT_IMAGE_TOOL=docker ../../../target/debug/pubsys --infra-config-path Infra.toml publish-kit --kit-path build/kits/core-kit --vendor bottlerocket --version v1.0.0 --build-id c232d4f-dirty
```

```
$ TWOLITER_KIT_IMAGE_TOOL=crane ../../../target/debug/pubsys --infra-config-path Infra.toml publish-kit --kit-path build/kits/core-kit --vendor bottlerocket --version v1.0.0 --build-id c232d4f-dirty
```

All of the above pushed as expected, and left the system in the correct state afterwards (i.e. cleaned up tempdirs and unnecessary docker images)

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Review by jmt-lab — APPROVED** [review]
LGTM

File: tools/buildsys/Cargo.toml [hunk 1]
```diff
@@ -20,7 +20,6 @@ pipesys = { version = "0.1", path = "../pipesys" }
 rand = { version = "0.8", default-features = false, features = ["std", "std_rng"] }
 regex = "1"
 reqwest = { version = "0.11", default-features = false, features = ["rustls-tls", "blocking"] }
-semver = { version = "1", features = ["serde"]}
```
Comment (bcressey) [inline tools/buildsys/Cargo.toml:23]: nit: could separate all these into a cleanup commit (but it's a good change and fine to include in this PR)

Might be good to add a `cargo machete` lint to CI.

File: tools/pubsys/src/kit/publish_kit/mod.rs [hunk 2]
```diff
@@ -26,52 +25,36 @@ pub(crate) struct PublishKitArgs {
...
-        );
-    }};
-}
+/*
+    async fn clean(&self, uri: &str) -> Result<()> {
+        self.cli.run(["manifest", "rm", uri]).await?;
+
+        Ok(())
+    }
+*/
```
Comment (bcressey) [inline tools/pubsys/src/kit/publish_kit/mod.rs]: delete?

File: tools/oci-cli-wrapper/src/crane.rs [hunk 3]
```diff
@@ -43,4 +48,46 @@ impl ImageTool for CraneCLI {
...
+
+    async fn push_oci_archive(&self, path: &Path, uri: &str) -> Result<()> {
+        let temp_dir = TempDir::new_in(path.parent().unwrap()).context(error::CraneTempSnafu)?;
+
+        let mut oci_file = File::open(path).context(error::ArchiveReadSnafu)?;
+
+        let mut oci_archive = TarArchive::new(&mut oci_file);
+        oci_archive
+            .unpack(temp_dir.path())
+            .context(error::ArchiveExtractSnafu)?;
```
Comment (bcressey) [inline tools/oci-cli-wrapper/src/crane.rs:60]: How much time does this step take? It's sort of nice to have just the tar archive as the output file from `rpm2kit`, but if it's a meaningful speed-up, we could have the unpacked contents be the "preferred" form and create the tar archive as a temporary file in the (already slow) Docker path.

File: tools/pubsys/src/kit/publish_kit/mod.rs [hunk 4]
```diff
@@ -26,52 +25,36 @@ pub(crate) struct PublishKitArgs {
...
-            result.status.success(),
-            error::PublishFailSnafu {
-                reason: String::from_utf8_lossy(&result.stderr)
-            }
-        );
-        result.stdout
-    }};
+    /// The image tool to use when pushing the kit. Valid arguments are docker and crane.
+    #[arg(long)]
+    image_tool: Option<String>,
```
Comment (bcressey) [inline tools/pubsys/src/kit/publish_kit/mod.rs]: Does this get used? Seems like we just call the `image_tool` helper function.

Comment (sam-berning) [reply tools/pubsys/src/kit/publish_kit/mod.rs]: whoops, this was a remnant from my initial implementation of `crane` support before #305. did not do a good job of cleaning up this file 🤦

File: tools/oci-cli-wrapper/src/crane.rs (hunk 3 above)
Comment (sam-berning) [reply tools/oci-cli-wrapper/src/crane.rs:60]: tested this with just the `tar` command against a recent build of the core kit. unpack took 16 seconds, and recompressing the contents took 1.5 seconds - so could definitely be an optimization worth taking. `docker` would take even longer to push an image but 1.5 seconds is pretty insignificant comparatively

**Review by bcressey — APPROVED** [review]

**Review by jmt-lab — APPROVED** [review]
LGTM

**Pull request merged**